### PR TITLE
Add scanned products to Order

### LINKED ISSUE #9653: [SKU Scanner] Add scanned product to order

### LINKED ISSUE #9654: [SKU Scanner] Map SKUs to Product IDs

### PR DESCRIPTION
**Pull request by iamgabrielma**
Part of https://github.com/woocommerce/woocommerce-ios/issues/9658

Closes: #9653 
Closes: #9654 

## Description
This PR deals with adding scanned products to an Order, by sending the decoded SKU to the store and retuning either a match (the SKU belongs to an existing product), or an error. 

Some caveats:
- Error handling will be dealt on a different PR, for now we just return a message to the console and do nothing.
- Only products can be added at the moment (not variations).
- Scanning the same product twice will add it twice (as a different row), we need to discuss if this should just increase the existing quantity or keep it as it is.

## Changes
* Upon tapping the Order's view button, we call the `addScannedProductToOrder` method
* This will attempt to find a matching product SKU in the store, via the existing `retrieveFirstProductMatchFromSKU` action 
  * If finds a match, passes the product ID to the OrderSynchronizer and updates the Order with the product
  * If doesn't find a match, returns a `.productNotFound` error

## Testing instructions
* Have a product with a decoded SKU in your store:
  * The easiest way to add the decoded barcode to a product is going to your store via wp-admin > WooCommerce > Products > Edit a product (not a variation) > Add the following SKU: `5366554338033`
  * An alternative is to create one in [this website](https://www.barcode-generator.org/) > open the app > Products > Select a product > Inventory > Tap on the barcode icon > scan the barcode created previously > tap done > tap save.
* On a physical device, go to Orders > `+` > `+ Add Products via SKU scanner` and scan the following QR:

See that does not add any product to the Order, and in the console we can read:

```
🍉 Detected SKU: long-sleeve vintage cotton t-shirt. Attempting to add to Order...
🍉 Failed to retrieve productID from sku long-sleeve vintage cotton t-shirt
```

4. Now scan again the following QR if your product uses `5366554338033`, or your own barcode/QR if you have created your own:

See that the product is added to the Order, and in the console we can read:

```
🍉 Detected SKU: 5366554338033. Attempting to add to Order...
🍉 Success: retrieved product ID 21 from sku 5366554338033
```

### CODE REVIEW

File: WooCommerce/Classes/ViewRelated/Orders/Order Creation/EditableOrderViewModel.swift [hunk 1]
```diff
@@ -1255,6 +1266,42 @@ extension EditableOrderViewModel {
...
+        }
+    }
+
+    /// Attempts to map SKU to product ID
+    ///
+    private func mapFromSKUtoProductID(sku: String, onCompletion: @escaping (Result<Int64, Error>) -> Void) {
+        let action = ProductAction.retrieveFirstProductMatchFromSKU(siteID: siteID, sku: sku, onCompletion: { result in
+            switch result {
+            case let .success(product):
+                print("🍉 Success: retrieved product ID \(product.productID) from sku \(sku)")
```
Comment (iamgabrielma) [inline WooCommerce/Classes/ViewRelated/Orders/Order Creation/EditableOrderViewModel.swift]: I'll remove the prints and watermelons before merging, just there to help with testing/debugging 🍉

File: WooCommerce/Classes/ViewRelated/Orders/Order Creation/EditableOrderViewModel.swift [hunk 2]
```diff
@@ -1255,6 +1266,42 @@ extension EditableOrderViewModel {
...
+    func addScannedProductToOrder(barcode sku: String?, onCompletion: @escaping (Result<Void, Error>) -> Void) {
+        guard let sku = sku else {
+            return onCompletion(.failure(ScannerError.nilSKU))
+        }
+
+        mapFromSKUtoProductID(sku: sku) { [weak self] result in
+            guard let self = self else { return }
+            switch result {
+            case let .success(productID):
+                self.configureProductRowViewModels()
```
Comment (joshheald) [inline WooCommerce/Classes/ViewRelated/Orders/Order Creation/EditableOrderViewModel.swift]: What's this call for? Is it only required for tests? It seems odd to do it before adding the product, and I'd kind of expect it to be in the completion chain from setting the product on the order...

File: WooCommerce/Classes/ViewRelated/Orders/Order Creation/EditableOrderViewModel.swift [hunk 3]
```diff
@@ -1255,6 +1266,42 @@ extension EditableOrderViewModel {
...
+                self.orderSynchronizer.setProduct.send(.init(product: .productID(productID), quantity: 1))
```
Comment (joshheald) [inline WooCommerce/Classes/ViewRelated/Orders/Order Creation/EditableOrderViewModel.swift:1281]: A note for a future PR perhaps – it _looks like_ this would set a quantity back to 1, if the same product was scanned a second time. What it actually does is add the product line again, with a quantity of 1. 

Perhaps instead we should increment the quantity of the existing product item by 1? That allows for a more traditional checkout-like use of the scanner, and keeps the line items list sensible.

File: WooCommerce/Classes/ViewRelated/Orders/Order Creation/EditableOrderViewModel.swift [hunk 4]
```diff
@@ -1255,6 +1266,42 @@ extension EditableOrderViewModel {
...
+    ///
+    private func mapFromSKUtoProductID(sku: String, onCompletion: @escaping (Result<Int64, Error>) -> Void) {
+        let action = ProductAction.retrieveFirstProductMatchFromSKU(siteID: siteID, sku: sku, onCompletion: { result in
+            switch result {
+            case let .success(product):
+                print("🍉 Success: retrieved product ID \(product.productID) from sku \(sku)")
+                onCompletion(.success(product.productID))
+            case .failure:
+                DDLogError("🍉 Failed to retrieve productID from sku \(sku)")
+                onCompletion(.failure(NSError()))
```
Comment (joshheald) [inline WooCommerce/Classes/ViewRelated/Orders/Order Creation/EditableOrderViewModel.swift]: This causes a one-time console error: 
```
-[NSError init] called; this results in an invalid NSError instance. It will raise an exception in a future release. Please call errorWithDomain:code:userInfo: or initWithDomain:code:userInfo:. This message shown only once.
```
I'm sure this won't be permanent anyway but it's best to resolve it now 😊

**Review by joshheald — APPROVED** [review]
I have a question, a couple of suggestions 🎶

Works as described – I did it with a UPC barcode too to check that.

File: WooCommerce/Classes/ViewRelated/Orders/Order Creation/EditableOrderViewModel.swift (hunk 1 above)
Comment (iamgabrielma) [reply WooCommerce/Classes/ViewRelated/Orders/Order Creation/EditableOrderViewModel.swift]: Removed on b4c49b16ed8a4e8ac3095610719c0171a916661b

File: WooCommerce/Classes/ViewRelated/Orders/Order Creation/EditableOrderViewModel.swift (hunk 4 above)
Comment (iamgabrielma) [reply WooCommerce/Classes/ViewRelated/Orders/Order Creation/EditableOrderViewModel.swift]: Definitely! Changed: 74025e4bc0ab6b9d46ba6577b424591972a2ca95

File: WooCommerce/Classes/ViewRelated/Orders/Order Creation/EditableOrderViewModel.swift (hunk 3 above)
Comment (iamgabrielma) [reply WooCommerce/Classes/ViewRelated/Orders/Order Creation/EditableOrderViewModel.swift:1281]: Yes, I initially planned to add a check to see if the product already exist in the order, and increase its quantity if it does, rather than sending it again to the synchronizer, however as we're gravitating to use this feature only when creating new orders for now ( pecCkj-F1#comment-308-p2 ) I left it as it is for now as the quantity will always be 1.

File: WooCommerce/Classes/ViewRelated/Orders/Order Creation/EditableOrderViewModel.swift (hunk 2 above)
Comment (iamgabrielma) [reply WooCommerce/Classes/ViewRelated/Orders/Order Creation/EditableOrderViewModel.swift]: > What's this call for? Is it only required for tests?  

TLDR;  As we're testing the contents of the ordersynchronizer now via the `currentOrderItems` computed var, I was able to remove this call on 350dc7e9a90d7c031c0e15e5f3da3e7583f06ef9 and still test the success case. This leaves us with a getter in production that is only used for testing, but IMHO is much better than having a random call in the scanner implementation, or to test productRows count.

Longer explanation: 

That's a good question, it took me quite a bit to understand what is happening, from what I understand:

Up until now we relied on counting the viewmodel's product rows when testing the EditableOrderViewModel. This made sense because the underlying logic for interacting with products relies on tapping them, which alters the state and triggers the different observers, updates state, etc, ... however by sending the products directly to the remote we seem to have some race condition in tests (perhaps also related the way we use MockStorage or we inject fake products in our test double) when we're expecting the updated values from the synchronizer's publishers and the contents of results controller: The notification seem to be sent before the variable itself is updated, as (the latest value in .sink is in willSet, not didSet), so we update the related methods by passing the out-of-date value.

This means we're late and the specific `ProductRowViewModel` fails to be created due to its coupling with fetching products from storage, as at the point we succeed to make our scanner call we notify the publisher but "we have not fetched yet" (we're accessing the old value, rather than the new), so the internal `allProducts` array is empty, which makes us return early from the different state that needs to be set for generating product rows. 

This call assured we're fetching from storage and the ProductRowViewmodels are set before we return from scanning. Since now we're not testing for product rows being rendered, but we test the contents of the sync order instead, we have removed this method call.

Comment (joshheald) [reply WooCommerce/Classes/ViewRelated/Orders/Order Creation/EditableOrderViewModel.swift]: Thanks for the explanation, and I agree that while the getter for tests only isn't brilliant, it's better than the alternative here!
